internal/graph/check_test.go

```diff
@@ -50,29 +51,125 @@ type document
+		resp, err = checker.ResolveCheck(ctx, &ResolveCheckRequest{
+			StoreID:            storeID,
+			TupleKey:           tuple.NewTupleKey("document:2", "editor", "user:x"),
+			ResolutionMetadata: &ResolutionMetadata{Depth: 2},
+		})
+		require.ErrorIs(t, err, ErrResolutionDepthExceeded)
+		require.Nil(t, resp)
 	})
-	require.NoError(t, err)
-	require.True(t, resp.Allowed)
 
-	resp, err = checker.ResolveCheck(ctx, &ResolveCheckRequest{
-		StoreID:            storeID,
-		TupleKey:           tuple.NewTupleKey("document:2", "editor", "user:x"),
-		ResolutionMetadata: &ResolutionMetadata{Depth: 2},
+	t.Run("exclusion_resolves_deterministically_1", func(t *testing.T) {
+		ds := memory.New()
+
+		storeID := ulid.Make().String()
+
+		err := ds.Write(context.Background(), storeID, nil, []*openfgav1.TupleKey{
+			tuple.NewTupleKey("document:budget", "admin", "user:*"),
+			tuple.NewTupleKeyWithCondition("document:budget", "viewer", "user:maria", "condX", nil),
+		})
+		require.NoError(t, err)
+
+		typedefs := parser.MustTransformDSLToProto(`model
+	schema 1.1
+type user
+
+type document
+  relations
+	define admin: [user:*]
+	define viewer: [user with condX] but not admin
+
+condition condX(x: int) {
+	x < 100
+}
+`).TypeDefinitions
+
+		checker := NewLocalChecker(ds)
+
+		ctx := typesystem.ContextWithTypesystem(context.Background(), typesystem.New(
+			&openfgav1.AuthorizationModel{
+				Id:              ulid.Make().String(),
+				TypeDefinitions: typedefs,
+				SchemaVersion:   typesystem.SchemaVersion1_1,
+			},
+		))
+
+		for i := 0; i < 2000; i++ {
+			// subtract branch resolves to {allowed: true} even though the base branch
+			// results in an error. Outcome should be falsey, not an error.
+			resp, err := checker.ResolveCheck(ctx, &ResolveCheckRequest{
+				StoreID:            storeID,
+				TupleKey:           tuple.NewTupleKey("document:budget", "viewer", "user:maria"),
+				ResolutionMetadata: &ResolutionMetadata{Depth: defaultResolveNodeLimit},
+			})
+			require.NoError(t, err)
+			require.False(t, resp.GetAllowed())
+		}
+	})
+
+	t.Run("exclusion_resolves_deterministically_2", func(t *testing.T) {
+		ds := memory.New()
+
+		storeID := ulid.Make().String()
+
+		err := ds.Write(context.Background(), storeID, nil, []*openfgav1.TupleKey{
+			tuple.NewTupleKeyWithCondition("document:budget", "admin", "user:maria", "condX", nil),
+		})
+		require.NoError(t, err)
+
+		typedefs := parser.MustTransformDSLToProto(`model
+			schema 1.1
+		type user
+
+		type document
+		  relations
+			define admin: [user with condX]
+			define viewer: [user] but not admin
+
+		condition condX(x: int) {
+			x < 100
+		}
+		`).TypeDefinitions
+
+		checker := NewLocalChecker(ds)
+
+		ctx := typesystem.ContextWithTypesystem(context.Background(), typesystem.New(
+			&openfgav1.AuthorizationModel{
+				Id:              ulid.Make().String(),
+				TypeDefinitions: typedefs,
+				SchemaVersion:   typesystem.SchemaVersion1_1,
+			},
+		))
+
+		for i := 0; i < 2000; i++ {
```

Comment:
We should fine-tune this number so that you consistently hit the code that you changed at least one time per test run. In my machine, 2000 iterations is not enough.

You can also make these individual tests parallel

Reply:
> We should fine-tune this number so that you consistently hit the code that you changed at least one time per test run. In my machine, 2000 iterations is not enough.

How would you propose we do that? There isn't a definitively way to guarantee the probability. What was enough on your machine? 

> You can also make these tests parallel

👍

Reply:
> How would you propose we do that? There isn't a definitively way to guarantee the probability. What was enough on your machine?

Never mind, with the latest changes both lines changed get hit pretty frequently